lib/nfc/protocols/st25tb/st25tb_poller.c

```diff
@@ -70,24 +69,28 @@ static NfcCommand st25tb_poller_run(NfcGenericEvent event, void* context) {
     NfcCommand command = NfcCommandContinue;
 
     if(nfc_event->type == NfcEventTypePollerReady) {
-        if(instance->state != St25tbPollerStateActivated) {
-            St25tbError error = st25tb_poller_async_activate(instance, instance->data);
-
-            if(error == St25tbErrorNone) {
-                instance->st25tb_event.type = St25tbPollerEventTypeReady;
-                instance->st25tb_event_data.error = error;
-                command = instance->callback(instance->general_event, instance->context);
-            } else {
-                instance->st25tb_event.type = St25tbPollerEventTypeError;
-                instance->st25tb_event_data.error = error;
-                command = instance->callback(instance->general_event, instance->context);
-                // Add delay to switch context
-                furi_delay_ms(100);
-            }
+        St25tbError error = St25tbErrorNone;
+        instance->st25tb_event.type = St25tbPollerEventTypeReady;
+        switch(instance->state) {
+        case St25tbPollerStateIdle:
+            error = st25tb_poller_async_select(instance, NULL);
+            break;
+        case St25tbPollerStateSelected:
+            error = st25tb_poller_async_read(instance, instance->data);
+            break;
+        case St25tbPollerStateRead:
+            instance->st25tb_event.type = St25tbPollerEventTypeReadSuccessful;
+            break;
+        }
+        if(error == St25tbErrorNone) {
+            instance->st25tb_event_data.error = error;
+            command = instance->callback(instance->general_event, instance->context);
         } else {
-            instance->st25tb_event.type = St25tbPollerEventTypeReady;
-            instance->st25tb_event_data.error = St25tbErrorNone;
+            instance->st25tb_event.type = St25tbPollerEventTypeError;
+            instance->st25tb_event_data.error = error;
             command = instance->callback(instance->general_event, instance->context);
+            // Add delay to switch context
+            furi_delay_ms(100);
```

Comment:
I can see the problem here. 
On first NfcEventTypePollerReady event you will  be in St25tbPollerStateIdle state and enter callback with St25tbPollerEventTypeReady event.
On second run you will have St25tbPollerStateSelected state and also spawn St25tbPollerEventTypeReady event.
And only on third run you will be in St25tbPollerStateRead state and spawn St25tbPollerEventTypeReadSuccessful event.

Sure, the application will work, but I think it's incorrect logic. I see that you split your poller in 3 states. I suggest adding separate state handlers array and call them in run() function, like it's don in mf_ultralight_poller.c for example. This would be much easier to read and expand.

Reply:
Also, if you do that, it would be nice if "async" functions will not change poller state. 

Firstly, it may cause some problems if you use those functions in different places, in sync API for example. 

Secondly, it would be much easier to understand your state machine. 

You should just call "async" functions and set new poller state based on returned error code

Reply:
I'll try to explain better how the thing works:
An api consumer only sees one of two events:
- St25tbPollerEventTypeReady: an st25tb tag is available in field and has been selected
- St25tbPollerEventTypeReadSuccessful: a tag has been fully read

The first event only exists to accommodate the existence of the sync api, this way it doesn't need to wait for a full read at every call.

I pushed a couple of commits that clean up the poller api and state machine according to your suggestions.